ext/pcntl/pcntl.c

```diff
@@ -385,6 +403,41 @@ PHP_FUNCTION(pcntl_waitpid)
 }
 /* }}} */
 
+#ifdef HAVE_WAITID
+PHP_FUNCTION(pcntl_waitid)
+{
+	zend_long idtype = P_PID;
+	zend_long id = 0;
+	zval *user_siginfo = NULL;
+	zend_long options = WEXITED;
+	int success;
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+		Z_PARAM_LONG(idtype)
```

Comment:
There should be a check + ValueError for the valid type.

Reply:
Strict idtype validation is platform-specific.

Here are the idtype definitions for platforms I checked:

https://github.com/torvalds/linux/blob/master/include/uapi/linux/wait.h#L16
https://github.com/openbsd/src/blob/master/sys/sys/wait.h#L87
https://github.com/NetBSD/src/blob/trunk/sys/sys/idtype.h#L42
https://cgit.freebsd.org/src/tree/sys/sys/wait.h#n93
https://github.com/illumos/illumos-gate/blob/b41e2fb6b2658a36accc73b3b86d99375e700dba/usr/src/uts/common/sys/procset.h#L58

Let's take NetBSD as a particular platform where valid idtypes are -1 <= idtype <= 15, with an additional value _P_MAXIDTYPE = 0x7fffffff (the  highest 32-bit two's complement representable positive integer). -1 on NetBSD translates to current process effective PID/GID.

I see four possible design choices here:

1. Leave idtype validation to the underlying C library function. This is what I would do, because what constitutes valid idtypes is platform-specific.
2. Implement an (ifdef) case-switch of various platform-specific validations. This could be a burden for long term maintenance, as ports to new platforms are introduced, and even existing implementations can change.
3. Implement some sort of weak validation, e.g. idtype >= -1. The downside is that such validation not catch all invalid values, and such an assumption could become incorrect on some platform in the future.
4. Support only POSIX-defined idtypes, regardless of the platform. This is clean to implement but users might find it overly restrictive.
                                             
Thoughts and suggestions are welcome.

Reply:
We generally do option 2 in other extensions.

But having 3 is also fine, it does not need to necessarily catch *all* issues, but erroring early with a ValueError for obviously wrong values is good.

Reply:
Ok, added weak validation per option 3 for now.

Another option would be better to check for EINVAL in the function itself, and convert it to ValueError. Not sure if this is consistent with the surrounding parts of the API.

Reply:
There is also a question of API consistency, other functions in the family (`pcntl_wait` and `pcntl_waitid`) set errno and don't throw ValueErrors, so having `pcntl_waitid` also throw ValueErrors would not be consistent with the rest of the package.

@devnexen would love to hear your thoughts on API design consistency, perhaps there is context I'm missing, and also of the idea of redundant checks vs converting EINVAL errnos to ValueError throwables.

Reply:
in fact, we should strive for better so other existing functions should throw ValueError too in further improvements.

Reply:
Sounds good, agreed.

Changed code to throw a ValueError when invalid arguments are passed.